wallet-new/src/Cardano/Wallet/Kernel/ChainState.hs

```diff
@@ -0,0 +1,150 @@
+    , uiPositiveStake   :: !Coin
+    , uiNegativeStake   :: !Coin
+    } deriving (Eq, Show)
+
+data ChainState = ChainState {
+      -- | Available update (if any)
+      csUpdateInfo       :: !(Maybe UpdateInfo)
+
+      -- | Block version data
+    , csBlockVersionData :: !BlockVersionData
+    }
+
+{-------------------------------------------------------------------------------
+  Updates
+-------------------------------------------------------------------------------}
+
+-- | Get the latest chain state
+getChainState :: MonadDBReadAdaptor IO -> IO ChainState
+getChainState rocksDB = withMonadDBRead rocksDB $ do
+    bvd       <- getAdoptedBVData
+    proposals <- getConfirmedProposals (Just $ svNumber curSoftwareVersion)
+    return ChainState{
+        csBlockVersionData = bvd
+      , csUpdateInfo       = fromProposals bvd proposals
+      }
+
+{-------------------------------------------------------------------------------
+  Getters
+-------------------------------------------------------------------------------}
+
+-- | Maximum transaction size
+maxTxSize :: ChainState -> Byte
+maxTxSize = bvdMaxTxSize . csBlockVersionData
+
+{-------------------------------------------------------------------------------
+  Internal auxiliary
+-------------------------------------------------------------------------------}
+
+fromProposals :: BlockVersionData
+              -> [ConfirmedProposalState]
+              -> Maybe UpdateInfo
+fromProposals _   []        = Nothing
+fromProposals bvd proposals = Just
+                            . fromProposal bvd
+                            . maximumBy (comparing cpsToNumericVersion)
+                            $ proposals
+  where
+    cpsToNumericVersion :: ConfirmedProposalState -> NumSoftwareVersion
+    cpsToNumericVersion = svNumber . upSoftwareVersion . cpsUpdateProposal
+
+-- | Construct 'UpdateInfo' from a 'ConfiurmedProposalState'
+--
+-- NOTE: This is a straight-forward adaptation of 'toCUpdateInfo' in the
+-- old wallet.
+fromProposal :: BlockVersionData -> ConfirmedProposalState -> UpdateInfo
+fromProposal bvd ConfirmedProposalState{..} = UpdateInfo {
+      uiSoftwareVersion = upSoftwareVersion
+    , uiBlockVersion    = upBlockVersion
+    , uiScriptVersion   = fromMaybe (bvdScriptVersion bvd)
+                                    (bvmScriptVersion upBlockVersionMod)
+    , uiImplicit        = cpsImplicit
+    , uiVotesFor        = votesFor
+    , uiVotesAgainst    = votesAgainst
+    , uiPositiveStake   = cpsPositiveStake
+    , uiNegativeStake   = cpsNegativeStake
+    }
+  where
+    UnsafeUpdateProposal {..} = cpsUpdateProposal
+    (votesFor, votesAgainst)  = countVotes cpsVotes
+
+-- | Return counts of negative and positive votes
+countVotes :: StakeholderVotes -> (Int, Int)
+countVotes = foldl' counter (0, 0)
+  where counter (n, m) vote = if isPositiveVote vote
+                              then (n + 1, m)
+                              else (n, m + 1)
```

Comment:
I just realised that using `foldl'` might not be enough here, because the tuple will be forced to WHNF, yes, but not its content 😉 

In practice this might not be a problem, but if we have lots and lots of votes, this will leak memory. Perhaps we should bang two exclamation marks in front of `n` and `m`?

Reply:
I was going to say "this is a direct copy and paste", but then I realized the follow up would be ".. from the old wallet" and I realized "where _there's_ your problem" :grin: Will fix.